apps/rush-lib/src/logic/DependencyAnalyzer.ts

```diff
@@ -0,0 +1,192 @@
+        //    (Even if it's the only version specifier anywhere in the repo, we still ignore it, because
+        //    otherwise the rule would be difficult to explain.)
+        if (alternativesForThisDependency.indexOf(dependencyVersion) > 0) {
+          ignoreVersion = true;
+        } else {
+          // Is it a local project?
+          const localProject: RushConfigurationProject | undefined =
+            this._rushConfiguration.getProjectByName(dependencyName);
+          if (localProject) {
+            // 2. If it's a symlinked local project, then it's not a candidate, because the package manager will
+            //    never even see it.
+            // However there are two ways that a local project can NOT be symlinked:
+            // - if the local project doesn't satisfy the referenced semver specifier; OR
+            // - if the local project was specified in "cyclicDependencyProjects" in rush.json
+            if (
+              !project.cyclicDependencyProjects.has(dependencyName) &&
+              semver.satisfies(localProject.packageJsonEditor.version, dependencyVersion)
+            ) {
+              ignoreVersion = true;
+            }
+          }
+
+          if (!ignoreVersion) {
+            let versionForDependency: Set<string> | undefined = allVersionsByPackageName.get(dependencyName);
+            if (!versionForDependency) {
+              versionForDependency = new Set<string>();
+              allVersionsByPackageName.set(dependencyName, versionForDependency);
+            }
+            versionForDependency.add(dependencyVersion);
+          }
+        }
+      }
+    }
+
+    return allVersionsByPackageName;
+  }
+
+  /**
+   * Generates the {@link IDependencyAnalysis.implicitlyPreferredVersionByPackageName} map,
+   * given the {@link IDependencyAnalysis.allVersionsByPackageName} map.
+   *
+   * @remarks
+   * The result of this function is not cached.
+   */
+  private _getImplicitlyPreferredVersionByPackageName(
+    commonVersionsConfiguration: CommonVersionsConfiguration,
+    allVersionsByPackageName: Map<string, Set<string>>
+  ): Map<string, string> {
+    // Only generate implicitly preferred versions for variants that request it
+    const useImplicitlyPreferredVersions: boolean =
+      commonVersionsConfiguration.implicitlyPreferredVersions !== undefined
+        ? commonVersionsConfiguration.implicitlyPreferredVersions
+        : true;
+
+    if (useImplicitlyPreferredVersions) {
+      // If any dependency has more than one version, then filter it out (since we don't know which version
+      // should be preferred).  What remains will be the list of preferred dependencies.
+      // dependency --> version specifier
+      const implicitlyPreferred: Map<string, string> = new Map<string, string>();
+      for (const [dep, versions] of allVersionsByPackageName) {
+        if (versions.size === 1) {
+          const version: string = Array.from(versions)[0];
```

Comment:
The old way of doing this using `versions[Symbol.iterator]().next().value` was better for performance and GC pressure. Creating an array just to read the first entry creates a throwaway array and requires calling next() at least twice.

Reply:
That's how I originally had it written, but Pete argued it was less readable.